MRI_geom/spd_fmri.py

```diff
@@ -0,0 +1,185 @@
+
+
+def fit_kernel_cv(log_euclidean_distance, labels,
+                  sigma=None, verbose=False):
+    kf = KFold(n_splits=10)
+    perf = {}
+    conf = {}
+    k = 0
+    if sigma is not None:
+        distance = np.exp(- np.square(log_euclidean_distance)/(sigma**2))
+    else:
+        distance = log_euclidean_distance
+    for train_index, test_index in kf.split(range(NB_GRAPHS)):
+        train_index = np.array(train_index)
+        test_index = np.array(test_index)
+        x = np.array(distance[train_index, :])[:, train_index]
+        x_test = np.array(distance[test_index, :])[:, train_index]
+        clf = SVC(kernel='precomputed')
+        clf.fit(x, labels[train_index])
+        y_train = clf.predict(x)
+        if verbose:
+            print("Training accuracy:",
+                  accuracy_score(labels[train_index], y_train))
+        y_test = clf.predict(x_test)
+        perf[k] = {'acc': accuracy_score(labels[test_index], y_test),
+                   'prec': precision_score(labels[test_index], y_test),
+                   'f1': f1_score(labels[test_index], y_test),
+                   'recall': recall_score(labels[test_index], y_test)}
+        conf[k] = confusion_matrix(labels[test_index], y_test)
+        k += 1
+        return perf, conf
+
+
+def main():
+    gamma = GAMMA
+    time_alg = {}
+    graphs, all_graphs, labels, stop, index_train = import_data()
+    distance = {option: np.zeros((NB_GRAPHS, NB_GRAPHS))
+                for option in DISTANCE}
+    tic = time.time()
+    for option in DISTANCE:
+        tic = time.time()
+        for c, g in enumerate(all_graphs):
+            hat_l1 = laplacian(g) + gamma*np.eye(DIM_SPACE)
+            lambda2, u = np.linalg.eigh(hat_l1)
+            inv_l1 = u.dot(np.diag(1.0/np.sqrt(lambda2)).dot(u.T))
+            for j in range(c):
+                hat_l2 = laplacian(all_graphs[j]) + gamma*np.eye(DIM_SPACE)
+                if option == "log_euclidean":
+                    distance[option][c, j] = np.linalg.norm(spd_space.group_log(hat_l1).squeeze(0)
+                                                            - spd_space.group_log(hat_l2).squeeze(0),
+                                                            'fro')
+                elif option == "frobenius":
+                    distance[option][c, j] = np.linalg.norm(hat_l1
+                                                            - hat_l2, 'fro')
+                else:
+                    distance[option][c, j] = np.linalg.norm(
+                                                            inv_l1.dot(spd_space.group_log(hat_l2).squeeze(0).dot(inv_l1)),
+                                                            'fro')
+        distance[option] = distance[option] + distance[option].T
+        toc = time.time()
+        time_alg[option] = toc-tic
+    print("Done computing distances")
+
+    print("Starting SVM fitting with Cross Validation")
+    sigmas = SIGMAS
+    mean_acc = {}
+    for option in DISTANCE:
```

Comment:
you need to split this fit kernel in multiple functions. In general try to avoid functions that are more than 20 lines of code. Function should fit in a decently sized terminal or it becomes hard to read/debug